src/Services/Finder.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace LaravelEnso\Upgrade\Services;
+
+use Illuminate\Support\Collection;
+use Illuminate\Support\Facades\File;
+use LaravelEnso\Upgrade\Contracts\MigratesStructure;
+use LaravelEnso\Upgrade\Contracts\Upgrade;
+use ReflectionClass;
+
+class Finder
+{
+    public static $dirs = ['.'];
+    public static $vendors = ['laravel-enso'];
+
+    public function upgrades(): Collection
+    {
+        return $this->initUpgrades()
+            ->filter->hasUpgrade()
+            ->map(fn ($dir) => $this->upgradeClasses($dir))
+            ->flatten();
+    }
+
+    private function initUpgrades(): Collection
+    {
+        return (new Collection(static::$vendors))
+            ->map(fn ($vendor) => base_path('vendor'.DIRECTORY_SEPARATOR.$vendor))
+            ->map(fn ($vendor) => File::directories($vendor))
+            ->flatten()
+            ->concat((new Collection(static::$dirs))->map(fn ($dir) => base_path($dir)))
+            ->map(fn ($path) => new Package($path))
+            ->filter->isPackage();
+    }
+
+    private function upgradeClasses(Package $package): Collection
+    {
+        return $package->upgradeClasses()
+            ->filter(fn ($class) => $this->isUpgrade($class))
+            ->map(fn ($class) => new $class)
+            ->map(fn ($upgrade) => $upgrade instanceof MigratesStructure
+                ? new Structure($upgrade)
+                : new $upgrade);
+    }
+
+    private function isUpgrade($class): bool
+    {
+        $reflection = new ReflectionClass($class);
```

Comment:
we can do without using the reflection class and use is_a($class, MigratesStructure::class, true) as it's shorter

Reply:
We don't have an object Here; therefore, I couldn't use `is_a` or `instanceof`. I didn't create an object from class because it would waste memory, and maybe it would have some side effects.

Reply:
`is_a` with the `true` flag also works with classes, in the latest version of php

Reply:
I think we should always use reflection when we can't use instanceof, at least for consistency and readability.

Apparently "is_a" won't resolve the "use" statement, we would have to give it the full q name.

Using it with a second true flag is more magic and less expressive.

At some point this function was deprecated then the deprecation was removed.